### PR TITLE
Flink Support for TIMESTAMP_NANOS

### PR DESCRIPTION
**Pull request by rodmeneses**
Supports the new TIMESTAMP_NANOS (v3 format) for Flink, including watermark generation

### CODE REVIEW

**Comment by pvary** [comment]
Do the schema conversions handle the nano ts correctly? Do we get a `Timestamp(9)` when we convert to Flink schema, or a nano timestamp when we convert from the Iceberg schema?

**Comment by rodmeneses** [comment]
> Do the schema conversions handle the nano ts correctly? Do we get a `Timestamp(9)` when we convert to Flink schema, or a nano timestamp when we convert from the Iceberg schema?

👍🏻

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java [hunk 1]
```diff
@@ -113,6 +113,15 @@ public LogicalType primitive(Type.PrimitiveType primitive) {
           // MICROS
           return new TimestampType(6);
         }
+      case TIMESTAMP_NANO:
+        Types.TimestampNanoType timestampNano = (Types.TimestampNanoType) primitive;
+        if (timestampNano.shouldAdjustToUTC()) {
+          // NANOS
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:119]: Why is this comment?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java [hunk 2]
```diff
@@ -411,6 +416,43 @@ public DecimalData read(DecimalData ignored) {
...
+  private static class NanosToTimestampReader
+      extends ParquetValueReaders.UnboxedReader<TimestampData> {
+    NanosToTimestampReader(ColumnDescriptor desc) {
+      super(desc);
+    }
+
+    @Override
+    public TimestampData read(TimestampData ignored) {
+      long value = readLong();
+      return TimestampData.fromInstant(Instant.ofEpochSecond(0, value));
```
Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:428]: How will this store nanoseconds?
How does the spark writer writes the nanoseconds?

Comment (pvary) [inline flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:428]: How could we cross check, that the Spark write/Flink read and Flink write/Spark read is correctly working?
Are we sure that this is the expected data format?

**Comment by pvary** [comment]
There are several tests which are using `DataGenerator` to generate test data - do we want to add nanos to them?

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java (hunk 1 above)
Comment (rodmeneses) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/TypeToFlinkType.java:119]: to be consistent with current comments already present for `MICROS`, ie:
```java
 if (timestamp.shouldAdjustToUTC()) {
          // MICROS
          return new LocalZonedTimestampType(6);
        } else {
          // MICROS
          return new TimestampType(6);
        }
```

File: flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java (hunk 2 above)
Comment (pvary) [reply flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/data/FlinkParquetReaders.java:428]: Maybe the generic java writer tests?

**Pull request closed without merging**